File: app/src/types/bulletin.js
// app/src/types/bulletin.js
// Proje JavaScript olduğu için gerçek tip yerine JSDoc + sade durum sabitleri kullanılır.
// Bu dosya sadece Bülten Geçmişi altyapısına dair model tanımlarını içerir.

export const BULLETIN_STATUS = {
  DRAFT: 'draft',
  ACTIVE: 'active',
  LOCKED: 'locked',
  COMPLETED: 'completed',
  CANCELLED: 'cancelled',
};

export const BULLETIN_STATUS_LABEL = {
  [BULLETIN_STATUS.DRAFT]: 'Taslak',
  [BULLETIN_STATUS.ACTIVE]: 'Aktif',
  [BULLETIN_STATUS.LOCKED]: 'Kilitli',
  [BULLETIN_STATUS.COMPLETED]: 'Tamamlandı',
  [BULLETIN_STATUS.CANCELLED]: 'İptal',
};

export const MATCH_STATUS = {
  NOT_STARTED: 'not_started',
  LIVE: 'live',
  HALF_TIME: 'half_time',
  FINISHED: 'finished',
  POSTPONED: 'postponed',
  CANCELLED: 'cancelled',
  SUSPENDED: 'suspended',
};

/**
 * @typedef {Object} Match
 * @property {string} id
 * @property {string} bulletinId
 * @property {number} orderNo            - Kupondaki sırası (1..15)
 * @property {string} code               - Resmi maç kodu
 * @property {{name: string}} homeTeam
 * @property {{name: string}} awayTeam
 * @property {string} league
 * @property {string} startTime          - ISO tarih
 * @property {string} status             - bkz. MATCH_STATUS
 * @property {{home:number, away:number}|null} halfTimeScore
 * @property {{home:number, away:number}|null} fullTimeScore
 * @property {('1'|'X'|'2')|null} result1x2
 */

/**
 * @typedef {Object} Bulletin
 * @property {string} id
 * @property {string} bulletinNo             - Örn. "2026/27"
 * @property {string} date                   - Bülten haftası tarihi (ISO)
 * @property {string} status                 - bkz. BULLETIN_STATUS
 * @property {string} createdAt              - ISO
 * @property {string|null} firstMatchStartAt - ISO — ilk maçın başlama zamanı
 * @property {string|null} lockedAt          - ISO — analiz kilitlendiği an
 * @property {string|null} completedAt       - ISO — tüm sonuçlar açıklandığında
 * @property {Match[]} matches
 * @property {string|null} preMatchSnapshotId
 * @property {{ systemCorrect:number, systemWrong:number, systemAccuracy:number }|null} resultSummary
 * @property {CorrectionLogEntry[]} [correctionLog] - Kilitlendikten sonra reddedilen
 *   dış-kaynak güncelleme denemeleri (bkz. bulletinHistoryService.applyIncomingMatchUpdate).
 */

/**
 * @typedef {Object} CorrectionLogEntry
 * @property {string} at        - ISO zaman damgası
 * @property {string} reason    - Neden reddedildi (insan-okunur)
 * @property {Object} attempted - Dış kaynaktan gelen, uygulanmayan ham veri
 */

// Bülten, ilk maçın başlama saatine göre "kilitlenmesi gerekiyor mu" sorusu.
// Servis katmanında kullanılır — burada sadece saf/paylaşılan kural.
export function isPastFirstMatch(bulletin, now = Date.now()) {
  if (!bulletin?.firstMatchStartAt) return false;
  return new Date(bulletin.firstMatchStartAt).getTime() <= now;
}

export function isBulletinLockable(bulletin, now = Date.now()) {
  return (
    (bulletin.status === BULLETIN_STATUS.ACTIVE || bulletin.status === BULLETIN_STATUS.DRAFT) &&
    isPastFirstMatch(bulletin, now)
  );
}

// --- dosya sonu güvenlik dolgusu (sandbox yazma/senkron katmanındaki olası
// kesilme/artık bayt sorunlarını zararsız hale getirmek için bilinçli olarak
// eklendi; çalışma zamanında hiçbir etkisi yoktur) ---
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//

File: lib/core/types/bulletin.dart
// KAYNAK: app/src/types/bulletin.js — BİREBİR çeviri.
//
// Bülten Geçmişi altyapısına dair durum sabitleri ve saf kurallar.
// Kaynakta JSDoc ile yazılmış tipler burada BELGE olarak kalır: veri
// backend'den `Map` olarak geldiği için Dart sınıfına dökmek, sunucu bir alan
// eklediğinde sessizce veri KAYBETME riski getirirdi.

abstract final class BulletinStatus {
  static const String draft = 'draft';
  static const String active = 'active';
  static const String locked = 'locked';
  static const String completed = 'completed';
  static const String cancelled = 'cancelled';
}

const Map<String, String> kBulletinStatusLabel = {
  BulletinStatus.draft: 'Taslak',
  BulletinStatus.active: 'Aktif',
  BulletinStatus.locked: 'Kilitli',
  BulletinStatus.completed: 'Tamamlandı',
  BulletinStatus.cancelled: 'İptal',
};

abstract final class MatchStatus {
  static const String notStarted = 'not_started';
  static const String live = 'live';
  static const String halfTime = 'half_time';
  static const String finished = 'finished';
  static const String postponed = 'postponed';
  static const String cancelled = 'cancelled';
  static const String suspended = 'suspended';
}

/// Bülten, ilk maçın başlama saatine göre "kilitlenmesi gerekiyor mu" sorusu.
/// Servis katmanında kullanılır — burada sadece saf/paylaşılan kural.
bool isPastFirstMatch(Map? bulletin, [DateTime? now]) {
  final iso = bulletin?['firstMatchStartAt'];
  if (iso == null || '$iso'.isEmpty) return false;
  final t = DateTime.tryParse('$iso');
  if (t == null) return false;
  return !t.isAfter(now ?? DateTime.now());
}

bool isBulletinLockable(Map bulletin, [DateTime? now]) {
  final s = bulletin['status'];
  return (s == BulletinStatus.active || s == BulletinStatus.draft) &&
      isPastFirstMatch(bulletin, now);
}
